hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java

```diff
@@ -1342,18 +1360,29 @@ private ClientMessage invokeWithPredicate(ClientMessage request, Predicate predi
         PagingPredicateImpl pagingPredicate = unwrapPagingPredicate(predicate);
         pagingPredicate.setIterationType(IterationType.VALUE);
 
-        ClientMessage request = MapValuesWithPagingPredicateCodec.encodeRequest(name, toData(predicate));
+        PagingPredicateHolder pagingPredicateHolder = PagingPredicateHolder.of(pagingPredicate, getSerializationService());
+        ClientMessage request = MapValuesWithPagingPredicateCodec.encodeRequest(name, pagingPredicateHolder);
+
         ClientMessage response = invokeWithPredicate(request, predicate);
         MapValuesWithPagingPredicateCodec.ResponseParameters resultParameters = MapValuesWithPagingPredicateCodec
                 .decodeResponse(response);
 
-        List<Entry> resultList = new ArrayList<>(resultParameters.response.size());
-        for (Entry<Data, Data> entry : resultParameters.response) {
-            K key = toObject(entry.getKey());
-            V value = toObject(entry.getValue());
-            resultList.add(new AbstractMap.SimpleImmutableEntry<>(key, value));
+        SerializationService serializationService = getSerializationService();
+
+        if (resultParameters.response.isEmpty()) {
+            return (Collection<V>) new ResultSet();
         }
-        return (Collection<V>) getSortedQueryResultSet(resultList, pagingPredicate, IterationType.VALUE);
+
+        int size = resultParameters.response.size();
+        List<Map.Entry> entries = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            K key = serializationService.toObject(resultParameters.response.get(i));
+            entries.add(new AbstractMap.SimpleImmutableEntry(key, null));
```

Comment:
Should we put the value instead here ?

Reply:
good catch (copy paste error). Corrected.